File: web/src/state/groups.ts
import { monotoneChainHull } from '../sim/convexHull';
import { segmentDistance, type Point } from '../sim/geometry';
import { polygonsOverlap, type Wall } from './model';

/**
 * Walls that touch or overlap, gathered into one group.
 *
 * Shapes drawn against each other should read as a single object, which is what
 * the dashed outline conveys. That used to be achieved by merging them into one
 * Wall, but merging cascades: draw anything against an enclosure and the
 * enclosure is swallowed. Grouping gives the same picture while every shape keeps
 * its own identity, colour and (eventually) deletability -- the grouping is
 * recomputed from scratch whenever walls change, so it can never accumulate.
 *
 * One group per connected set of shapes, one hull over all of it: a shape
 * touching nothing is a group of one and is hulled just the same, and a freehand
 * trace shapes the hull of whatever it touches like any other member -- a
 * squiggle laid across two buildings puts all three under one outline.
 *
 * Border frames are the one exception, and are skipped entirely: they get no
 * outline, and they join nothing to anything. A frame's hull is the solid
 * rectangle it encloses, so it would draw the room's walkable interior as if it
 * were the obstacle -- and since a frame reaches all the way round the map,
 * everything inside it touches it and would collapse into that single hull,
 * leaving the buildings inside with no outline of their own. Skipping the frame
 * is what lets two rectangles drawn inside an enclosure be hulled as the two
 * rectangles they are.
 */
export interface WallGroup {
  /** Every wall in the group, lowest id first. */
  wallIds: number[];
  /** Convex hull over every point of every wall in the group. */
  hull: Point[];
}

/** How close two walls must come to count as touching, in world units. */
export const GROUP_TOLERANCE = 1;

function wallsTouchWithin(a: Wall, b: Wall, tolerance: number): boolean {
  for (const pa of a.polygons) {
    for (const pb of b.polygons) {
      if (polygonsOverlap(pa, pb)) return true;
      for (let i = 0; i < pa.length; i++) {
        const a1 = pa[i];
        const a2 = pa[(i + 1) % pa.length];
        for (let j = 0; j < pb.length; j++) {
          const b1 = pb[j];
          const b2 = pb[(j + 1) % pb.length];
          if (segmentDistance(a1, a2, b1, b2) <= tolerance) return true;
        }
      }
    }
  }
  return false;
}

export function groupWalls(allWalls: Wall[], tolerance = GROUP_TOLERANCE): WallGroup[] {
  const walls = allWalls.filter((w) => !w.isBorder);
  const n = walls.length;
  if (n === 0) return [];

  // Union-find over touching pairs.
  const parent = Array.from({ length: n }, (_, i) => i);
  const find = (i: number): number => {
    let root = i;
    while (parent[root] !== root) root = parent[root];
    while (parent[i] !== root) { const next = parent[i]; parent[i] = root; i = next; }
    return root;
  };
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (find(i) === find(j)) continue;
      if (wallsTouchWithin(walls[i], walls[j], tolerance)) parent[find(i)] = find(j);
    }
  }

  const byRoot = new Map<number, number[]>();
  for (let i = 0; i < n; i++) {
    const root = find(i);
    const bucket = byRoot.get(root);
    if (bucket) bucket.push(i);
    else byRoot.set(root, [i]);
  }

  // Sorted so the lowest id is first: the outline's colour comes from the first
  // member, and must not change as unrelated shapes are added elsewhere.
  const byId = (a: number, b: number) => a - b;

  const groups: WallGroup[] = [];
  for (const members of byRoot.values()) {
    const points: Point[] = [];
    for (const i of members) points.push(...walls[i].polygons.flat());
    groups.push({
      wallIds: members.map((i) => walls[i].id).sort(byId),
      hull: monotoneChainHull(points),
    });
  }
  return groups.sort((a, b) => a.wallIds[0] - b.wallIds[0]);
}

File: ios/Sources/WalkyCore/Groups.swift
import Foundation

/// Walls that touch or overlap, gathered into one group. Ports `src/state/groups.ts`.
///
/// Shapes drawn against each other should read as a single object, which is
/// what the dashed outline conveys. That used to be achieved by merging them
/// into one `Wall`, but merging cascades: draw anything against an enclosure
/// and the enclosure is swallowed. Grouping gives the same picture while every
/// shape keeps its own identity -- and it is recomputed from scratch whenever
/// walls change, so unlike merging it can never accumulate.
///
/// Border frames are the one exception and are skipped entirely. A frame's hull
/// is the solid rectangle it encloses, so it would draw the room's walkable
/// interior as if it were the obstacle -- and since a frame reaches all the way
/// round the map, everything inside it touches it and would collapse into that
/// single hull.
public struct WallGroup {
  /// Every wall in the group, lowest id first.
  public var wallIds: [Int]
  /// Convex hull over every point of every wall in the group.
  public var hull: [Point]
}

/// How close two walls must come to count as touching, in world units.
public let GROUP_TOLERANCE: Double = 1

private func wallsTouchWithin(_ a: Wall, _ b: Wall, _ tolerance: Double) -> Bool {
  for pa in a.polygons {
    for pb in b.polygons {
      if polygonsOverlap(pa, pb) { return true }
      for i in 0..<pa.count {
        let a1 = pa[i]
        let a2 = pa[(i + 1) % pa.count]
        for j in 0..<pb.count {
          let b1 = pb[j]
          let b2 = pb[(j + 1) % pb.count]
          if segmentDistance(a1, a2, b1, b2) <= tolerance { return true }
        }
      }
    }
  }
  return false
}

public func groupWalls(_ allWalls: [Wall], _ tolerance: Double = GROUP_TOLERANCE) -> [WallGroup] {
  let walls = allWalls.filter { !$0.isBorder }
  let n = walls.count
  if n == 0 { return [] }

  // Union-find over touching pairs.
  var parent = Array(0..<n)
  func find(_ start: Int) -> Int {
    var root = start
    while parent[root] != root { root = parent[root] }
    var i = start
    while parent[i] != root { let next = parent[i]; parent[i] = root; i = next }
    return root
  }
  for i in 0..<n {
    for j in (i + 1)..<n {
      if find(i) == find(j) { continue }
      if wallsTouchWithin(walls[i], walls[j], tolerance) { parent[find(i)] = find(j) }
    }
  }

  // Insertion-ordered buckets: JS `Map` preserves the order roots are first
  // seen, and the groups are sorted by lowest id at the end anyway -- but the
  // hull points are gathered in member order, so the order matters to the
  // hull's input and therefore to nothing observable. Kept ordered regardless,
  // for the same reason Navigation.fieldOrder is.
  var rootOrder: [Int] = []
  var byRoot: [Int: [Int]] = [:]
  for i in 0..<n {
    let root = find(i)
    if byRoot[root] == nil { rootOrder.append(root); byRoot[root] = [] }
    byRoot[root]!.append(i)
  }

  var groups: [WallGroup] = []
  for root in rootOrder {
    let members = byRoot[root]!
    var points: [Point] = []
    for i in members { points.append(contentsOf: walls[i].polygons.flatMap { $0 }) }
    // Sorted so the lowest id is first: the outline's colour comes from the
    // first member, and must not change as unrelated shapes are added elsewhere.
    groups.append(WallGroup(wallIds: members.map { walls[$0].id }.sorted(),
                            hull: monotoneChainHull(points)))
  }
  return groups.sorted { $0.wallIds[0] < $1.wallIds[0] }
}
